h2o-automl/src/main/java/ai/h2o/automl/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,184 @@
+package ai.h2o.automl.targetencoding;
+
+import water.Iced;
+import water.MRTask;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+import water.util.DistributedException;
+import water.util.IcedHashMap;
+
+import java.util.Objects;
+
+class BroadcastJoinForTargetEncoder {
+  
+  static class CompositeLookupKey extends Iced {
+    // Note: Consider using indexes of levels instead of strings to save space. It will probably require 
```

Comment:
I agree with this comment. It would be more H2O-like approach to represent the values by their indices in the domain.

You also don't need this structure, you can have a Map per-fold instead of one big Map. When categoricals are represented as indices, this map can be implemented using a plain java array.

If the Vec is cached on the node, the domain doesn't have to be sent over the wire. Having a structure that doesn't need to represent the domain thus saves possibly a lot of memory for large domains.

Same applies to `-1` fold - if you have this structure per fold then the memory is used more efficiently and much less data is sent over the wire.

I would suggest making this change.

Reply:
@michalkurka I have changed representation and store everything in arrays now. Locally single/multinode tests are fine.... waiting for Jenkins. Could you please take a look at the implementation? The only thing that bothers me is a reduce phase when we have to merge arrays. It is O(m*2n), where `m` is number of categorical levels and `n` is number of folds. In most cases number of folds is not that big so it is probably fine.

Reply:
@michalkurka I've tested one stress scenario and time performance is almost the same. I'm going to run scenario that was causing OOM and see if approach with arrays improved situation.